nameserver/dns.go

```diff
@@ -1,117 +1,272 @@
+func (d *DNSServer) handleLocal(defaultMaxResponseSize int) func(dns.ResponseWriter, *dns.Msg) {
+	return func(w dns.ResponseWriter, req *dns.Msg) {
+		d.ns.debugf("local request: %+v", *req)
+		if len(req.Question) != 1 || req.Question[0].Qtype != dns.TypeA {
+			d.errorResponse(req, dns.RcodeNameError, w)
+			return
+		}
+
+		hostname := req.Question[0].Name
+		addrs := d.ns.Lookup(hostname)
+
+		response := dns.Msg{}
+		response.RecursionAvailable = true
+		response.Authoritative = true
+		response.SetReply(req)
+		response.Answer = make([]dns.RR, len(addrs))
+
+		header := dns.RR_Header{
+			Name:   hostname,
+			Rrtype: dns.TypeA,
+			Class:  dns.ClassINET,
+			Ttl:    d.ttl,
+		}
+
+		for i, addr := range addrs {
+			ip := addr.IP4()
+			response.Answer[i] = &dns.A{Hdr: header, A: ip}
+		}
+
+		shuffleAnswers(&response.Answer)
+		maxResponseSize := getMaxResponseSize(req, defaultMaxResponseSize)
+		truncateResponse(&response, maxResponseSize)
+
+		d.ns.debugf("response: %+v", response)
+		if err := w.WriteMsg(&response); err != nil {
+			d.ns.infof("error responding: %v", err)
+		}
 	}
-	return maxLen
 }
 
-func init() {
-	rand.Seed(time.Now().UTC().UnixNano())
+func (d *DNSServer) handleReverse(client *dns.Client, defaultMaxResponseSize int) func(dns.ResponseWriter, *dns.Msg) {
+	return func(w dns.ResponseWriter, req *dns.Msg) {
+		d.ns.debugf("reverse request: %+v", *req)
+		if len(req.Question) != 1 || req.Question[0].Qtype != dns.TypePTR {
+			d.errorResponse(req, dns.RcodeNameError, w)
+			return
+		}
+
+		ipStr := strings.TrimSuffix(req.Question[0].Name, "."+reverseDNSdomain)
+		ip, err := address.ParseIP(ipStr)
+		if err != nil {
+			d.errorResponse(req, dns.RcodeNameError, w)
+			return
+		}
+
+		hostname, err := d.ns.ReverseLookup(ip.Reverse())
+		if err != nil {
+			d.handleRecursive(client, defaultMaxResponseSize)(w, req)
+			return
+		}
+
+		response := dns.Msg{}
+		response.RecursionAvailable = true
+		response.Authoritative = true
+		response.SetReply(req)
+
+		header := dns.RR_Header{
+			Name:   req.Question[0].Name,
+			Rrtype: dns.TypePTR,
+			Class:  dns.ClassINET,
+			Ttl:    d.ttl,
+		}
+
+		response.Answer = []dns.RR{&dns.PTR{
+			Hdr: header,
+			Ptr: hostname,
+		}}
+
+		maxResponseSize := getMaxResponseSize(req, defaultMaxResponseSize)
+		truncateResponse(&response, maxResponseSize)
+
+		d.ns.debugf("response: %+v", response)
+		if err := w.WriteMsg(&response); err != nil {
+			d.ns.infof("error responding: %v", err)
+		}
+	}
 }
 
-// shuffleAnswers reorders answers for very basic load balancing
-func shuffleAnswers(answers []dns.RR) []dns.RR {
-	if len(answers) > 1 {
-		for i := range answers {
-			j := rand.Intn(i + 1)
-			answers[i], answers[j] = answers[j], answers[i]
+func (d *DNSServer) handleRecursive(client *dns.Client, defaultMaxResponseSize int) func(dns.ResponseWriter, *dns.Msg) {
+	return func(w dns.ResponseWriter, req *dns.Msg) {
+		d.ns.debugf("recursive cdrequest: %+v", *req)
```

Comment:
I'm not sure you can just send the same request we get from the client to the upstream server. This could lead to a conflict if two clients send a request with the same message ID. I would make a copy...